modules/gapi/src/streaming/onevpl/accelerators/dx11_alloc_resource.cpp

```diff
@@ -0,0 +1,404 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2021 Intel Corporation
+
+#ifdef HAVE_ONEVPL
+#include "streaming/onevpl/accelerators/dx11_alloc_resource.hpp"
+#include "streaming/onevpl/accelerators/utils/shared_lock.hpp"
+#include "logger.hpp"
+
+#ifdef HAVE_DIRECTX
+#ifdef HAVE_D3D11
+
+namespace cv {
+namespace gapi {
+namespace wip {
+namespace onevpl {
+
+LockAdapter::LockAdapter(mfxFrameAllocator origin_allocator) :
+    lockable_allocator(origin_allocator),
+    impl() {
+    GAPI_DbgAssert((lockable_allocator.Lock && lockable_allocator.Unlock) &&
+                   "Cannot create LockAdapter for empty origin allocator");
+
+    // abandon unusable c-allocator interfaces
+    // because LockAdapter requires Lock & Unlock only
+    lockable_allocator.Alloc = nullptr;
+    lockable_allocator.Free = nullptr;
+    lockable_allocator.pthis = nullptr;
+}
+
+size_t LockAdapter::read_lock(mfxMemId mid, mfxFrameData &data) {
+    size_t prev_lock_count = 0;
+    if (impl) {
+        prev_lock_count = impl->shared_lock();
+    }
+
+    // dispatch to VPL allocator using READ access mode
+    mfxStatus sts = MFX_ERR_LOCK_MEMORY;
+    try {
+        sts = lockable_allocator.Lock(nullptr, mid, &data);
+    } catch(...) {
+    }
+
+    // adapter will throw error if VPL frame allocator fails
```

Comment:
If it throws then you just caught and discarded it in the line above. Why just not handle it with RAII lock over `impl` here?

Reply:
yes, some ugly code below, but cover both cases for exception and regular fail